Community-Supported/clouddb-extractor/bigquery_extractor.py

```diff
@@ -0,0 +1,514 @@
+            table_name=TableName("Extract", hyper_table_name), columns=target_cols
+        )
+
+        return target_schema
+
+    def _query_to_hyper_files(self, sql_query, hyper_table_name="Extract"):
+        """
+        Executes sql_query against the source database and writes the output to one or more Hyper files
+        Returns a list of output Hyper files
+
+        sql_query -- SQL string to pass to the source database
+        hyper_table_name -- Name of the target Hyper table, default=Extract
+        """
+
+        # Dry run to estimate result size
+        job_config = bigquery.QueryJobConfig(dry_run=True, use_query_cache=False)
+        dryrun_query_job = bq_client.query(sql_query, job_config=job_config)
+        dryrun_bytes_estimate = dryrun_query_job.total_bytes_processed
+        logger.info("This query will process {} bytes.".format(dryrun_bytes_estimate))
+
+        if dryrun_bytes_estimate > MAX_QUERY_SIZE:
+            raise QuerySizeLimitError(
+                "This query will return than {MAX_QUERY_SIZE} bytes"
+            )
+
+        query_job = bq_client.query(sql_query)
+
+        # Determine table structure
+        query_temp_table = bq_client.get_table(query_job.destination)
+        target_table_def = self._hyper_table_definition(
+            query_temp_table, hyper_table_name
+        )
+
+        def query_job_iter():
+            return bq_client.list_rows(query_job.destination)
+
+        return self._query_result_to_hyper_files(query_job_iter, target_table_def)
+
+    def load_sample(
+        self,
+        source_table,
+        tab_ds_name,
+        sample_rows=SAMPLE_ROWS,
+        publish_mode=TSC.Server.PublishMode.CreateNew,
+    ):
+        """
+        Loads a sample of rows from source_table to Tableau Server
+
+        source_table (string): Source table ref ("project ID.dataset ID.table ID")
+        tab_ds_name (string): Target datasource name
+        publish_mode: One of TSC.Server.[Overwrite|CreateNew] (default=CreateNew)
+        sample_rows (int): How many rows to include in the sample (default=SAMPLE_ROWS)
+        """
+
+        # loads in the first sample_rows of a bigquery table
+        # set publish_mode=TSC.Server.PublishMode.Overwrite to refresh a sample
+        sql_query = "SELECT * FROM `{}` LIMIT {}".format(source_table, sample_rows)
+        output_hyper_files = self._query_to_hyper_files(sql_query, tab_ds_name)
+        first_chunk = True
+        for path_to_database in output_hyper_files:
+            if first_chunk:
+                self._publish_hyper_file(path_to_database, tab_ds_name, publish_mode)
+                first_chunk = False
+            else:
+                self._update_datasource_from_hyper_file(
+                    path_to_database=path_to_database,
+                    tab_ds_name=tab_ds_name,
+                    action="INSERT",
+                )
+            self._publish_hyper_file(path_to_database, tab_ds_name, publish_mode)
```

Comment:
why are we publishing the same Hyper file twice? Didn't we just publish in line 222?